Evilify geiser-doc-mode-map and add keybindings in scheme layer

dalanicolai opened the pull request:
In the scheme REPL, RET is bound to `geiser-repl--maybe-send` which most times
sends the input instead of inserting a newline. Therefore this separate
keybinding for that will always insert the newline is required.

Also the `geiser-doc-mode` is read-only and should get evilified to not shadow its default keybindings.

dalanicolai commented:
@lebensterben RET is already bound to such function (but so that function does not work very well). Anyway, I could change the new binding from `C-RET` to `S-RET` but I am not sure. Let us discuss a little...

I agree that the Ctrl should be on the inside of the Fn (as is the default on my Lenovo keyboard. I wonder how much thought/testing manufacturers that place the Ctrl on the outside put in that design). But usually those buttons can be swapped very easily via the bios or the operating system (which we should probably recommend to users).

So for me, pressing Shift or Ctrl really makes no difference and I would more think about consistency with other modes.
E.g. in the customize menu's we use `C-j` to add a newline (so that is Ctrl`). And I think (not sure) in most repl/notebook interfaces (ipython/jupyter/Mathematica) `S-RET` is used to send to REPL.

You have any more "references"? What should we decide?

lebensterben commented:
> And I think (not sure) in most repl/notebook interfaces (ipython/jupyter/Mathematica) `S-RET` is used to send to REPL.

No. For vast majority of applications S-enter is to start a new line.
https://defkey.com/what-means/shift-enter

While Command-enter or Control-enter usually do something completely different
https://defkey.com/what-means/control-enter
https://defkey.com/what-means/command-enter

dalanicolai commented:
Ah that's funny, I never worked with any of those programs, except with Jupyter and Spyder (and Mathematica which is not in the list), and indeed they send to REPL, but now I see those are exception (although maybe for a good reason).
Also I see in your list that `C-RET` also is mostly used to insert a line break, so I would not call that completely different (`S-RET` is as much used to do completely different things).

So I still think the choice is quite arbitrary. But I have no problem to swap them if you like (and break with Mathematica and Jupyter behavior)

dalanicolai commented:
@lebensterben I actually tried them first (I think I took the `g j` and `g k` from there), but then for some reason I preferred to add bindings myself. But I do not remember why I preferred that, because now those binding look quite acceptable.

However, those bindings do not add much to what we have already. And for sure I prefer to use evilified-state for read-only buffers. So we could selectively adopt the few useful extra keybindings they have. But until now I did not miss any of those bindings (still might be nice to add them). I will have a quick look again tomorrow, and maybe adopt some of them.

dalanicolai commented:
I've added some more bindings, should be fairly complete now

lebensterben commented on layers/+lang/scheme/funcs.el:
```diff
@@ -0,0 +1,4 @@
+(defun geiser-customize-implementation ()
```
what's this for?

dalanicolai replied:
Ah well, as up to now I only use very few geiser functions, I just used the `geiser` menu in the menu-bar to check out the available geiser functions in the different geiser modes.
This menu has an entry to jump to the customize menu for the currently used scheme implementation.
I thought that is might be a nice service, to provide a similar function under the `SPC m h` menu, but I am just trying it out.
I am happy to hear any positive or negative opinion about it, maybe I am becoming too user friendly :laughing:

lebensterben replied:
I think this is not really useful.
If you find particular customization variables that are commonly used on a per-file/per-project bases, you can make some toggle functions and bind them under `SPC m T`.

That provides easy access to particular, concrete functionalities.

lebensterben commented on layers/+lang/scheme/README.org:
```diff
@@ -183,15 +184,31 @@ When enabled the symbol =🆂= should be displayed in the mode-line.
+
+| Key binding | Description                      |
+|-------------+----------------------------------|
+| ~C-RET~     | geiser-repl--newline-and-indent  |
```
S-RET

lebensterben commented on layers/+lang/scheme/packages.el:
```diff
@@ -102,17 +102,44 @@
-        "q" 'geiser-repl-exit
+        "b" 'geiser-pop-symbol-stack
         "c" 'geiser-repl-clear-buffer
         "i" 'geiser-repl-interrupt
```
this one should be C-d.
Because, literally, ^D has the meaning of end of transmission.

And this is indeed the default key binding of most repl to signal EOT.

"i" is not really meaningful. I'd more likely to think it represents "insert".

Then insert lambda is "il", and import module is "im".

dalanicolai replied:
Yes, nice suggestions :+1:

dalanicolai replied:
It [seems that the `geiser-pop-symbol-stack` is quite useless in the REPL](https://www.nongnu.org/geiser/geiser_4.html#Jumping-around), I put it here because `evil-collection` has it. But I will remove it.

lebensterben replied:
```elisp
(defun geiser-edit-symbol-at-point (&optional arg)
  "Opens a new window visiting the definition of the symbol at point.
With prefix, asks for the symbol to edit.")

(defun geiser-pop-symbol-stack ()
  "Pop back to where \\[geiser-edit-symbol-at-point] was last invoked.")
```

I think both functions are not useful...
To jump to the symbol-at-point, you can do `g d` or `SPC m g d`. Then `C-o` brings you back to the original marker.

lebensterben commented on layers/+lang/scheme/packages.el:
```diff
@@ -102,17 +102,44 @@
       (spacemacs/set-leader-keys-for-major-mode 'geiser-repl-mode
-        "q" 'geiser-repl-exit
+        "b" 'geiser-pop-symbol-stack
         "c" 'geiser-repl-clear-buffer
```
This can be C-l.
^l is form feed and is the default binding for clearing current screen.

the current binding is understandable enough though.

lebensterben replied:
After thinking about this again, we probably should use more cubersome key binding than "c".

a single key is easy to press, and is easy to press by mistake.
Clearing the repl seems like a "destructive" action, so we'd better make the key binding idiot-proof.

"C-l" would be the standard binding for clearing screen, but you may use "C" ("s-c").

lebensterben commented on layers/+lang/scheme/packages.el:
```diff
@@ -102,17 +102,44 @@
+        "hd" 'geiser-doc-symbol-at-point
+        "s" 'geiser-squarify
+        "q" 'geiser-repl-exit
+        "hc" 'geiser-customize-implementation)
```
I don't recall we have added any binding for customizing a particular mode.

Is this a common part of workflow?

If you really want, something in ", T" is better because that's the prefix for toggles.

dalanicolai replied:
You mean `, T` for toggle squarify? Or to go to customize?

lebensterben replied:
no, I mean geiser customize implementation.

lebensterben replied:
it's just not conventional. We haven't added similar bindings to other major modes.
If `SPC m h c` is a good binding for customization, it should also be applied to other major modes to maintain consistency.

It's better to open a separate PR for this. We need to be more cautious when devising a new convention.

lebensterben commented on layers/+lang/scheme/packages.el:
```diff
@@ -102,17 +102,44 @@
         "u" 'geiser-repl-unload-function
-        "h" 'geiser-doc-symbol-at-point))))
+        "hd" 'geiser-doc-symbol-at-point
+        "s" 'geiser-squarify
```
Is this a common part of workflow?

Some other modes may also accomplish this functionality, eg in parinfer-rust-mode if you change the opening parenthesis to bracket it changes the closing one (or vice versa).

p.s. one real issue is the command name itself is really not informative...

dalanicolai replied:
Well it is `almost` provided by `sp-rewrap-sexp` or `evil-surround` (`c s ) ]`). But they both require you to type explicitly the new type of enclosing.
I know in racket and chez, brackets are equivalent to parenthesis (they are used interchangeably to "improve" readability. I don't know about scheme in general). So anyway, I guess this shortcut is quite handy.

lebensterben commented on layers/+lang/scheme/packages.el:
```diff
@@ -102,17 +102,44 @@
+        ">" 'geiser-doc-next
+        "<" 'geiser-doc-previous
+
+        "gb" 'geiser-doc-previous
```
for "gb" and "gf"

do you mean backward and forward?

normally we just use "gn" and "gp" as next and previous.

lebensterben commented on layers/+lang/scheme/packages.el:
```diff
@@ -102,17 +102,44 @@
+        "q" 'geiser-repl-exit
+        "hc" 'geiser-customize-implementation)
+
+      (evilified-state-evilify-map geiser-doc-mode-map
```
do you think we should also added bindings for forward-button and backward-button, as it does in evil-collections? (I've no strong preference on this one)

dalanicolai replied:
Well generally I use `o` for this, and otherwise personally I prefer `TAB/S-TAB` which are already the default emacs (i.e. evilified) bindings. But indeed I should at least add `TAB/S-TAB` to the docs. Or do you think I should add evil-collection's `C-j/C-k` also?

lebensterben replied:
C-j/C-k is almost as common as TAB/S-TAB in various evil bindings.
So you can add both.

dalanicolai commented:
I have updated the keybindings (implemented suggestions) and updated the docs

lebensterben commented on layers/+lang/scheme/README.org:
```diff
@@ -183,15 +184,41 @@ When enabled the symbol =🆂= should be displayed in the mode-line.
+| Key binding | Description           |
+|-------------+-----------------------|
+| ~S-RET~     | Insert newline indent |
+| ~C-l~       | Clear buffer          |
```
~C-l / C~ since you added both

lebensterben commented on layers/+lang/scheme/README.org:
```diff
@@ -183,15 +184,41 @@ When enabled the symbol =🆂= should be displayed in the mode-line.
+| ~g n~       | Next page             |
+| ~g d~       | Goto definition       |
+| ~g z~       | Switch to repl        |
+| ~TAB/S-TAB~ | Next/previous button  |
```
It's clearer if
```
| ~TAB/C-j~   | Next button     |
| ~S-TAB/C-k~ | Previous button |
```

lebensterben commented on layers/+lang/scheme/README.org:
```diff
@@ -183,15 +184,41 @@ When enabled the symbol =🆂= should be displayed in the mode-line.
+| ~SPC m i l~ | Insert lambda                                     |
+| ~SPC m i m~ | Import module                                     |
+| ~SPC m u~   | Unload function                                   |
+| ~SPC m s~   | toggle surrounding parenthesis <-> brackets       |
```
Now I think this is a good binding.
In visual state, `s` binds to `evil-surrond-edit`, which let you change/delete surrounding charater.
Analogously `SPC m s` toggles the surround () to/from [].

lebensterben reviewed (changes requested):
Overall good. Just one issue, is `geiser-customize-implementation` really important so we give it a dedicated keybinding "SPC m h c"

If certain variables are commonly set on per-file/per-session basis, it's better to make them toggles if possible, under "SPC m T" prefix.

dalanicolai commented:
I think I forgot to 'submit the review'...

dalanicolai commented:
> Overall good. Just one issue, is `geiser-customize-implementation` really important so we give it a dedicated keybinding "SPC m h c"
> 
> If certain variables are commonly set on per-file/per-session basis, it's better to make them toggles if possible, under "SPC m T" prefix.

No it is not important, but https://github.com/syl20bnr/spacemacs/pull/14758#discussion_r635157422 is what I commented in the 'review that I forgot to submit'. Anyway, it would be no problem to remove it...

dalanicolai commented on layers/+lang/scheme/README.org:
```diff
@@ -157,7 +158,7 @@ When enabled the symbol =🆂= should be displayed in the mode-line.
 | ~SPC m m e~ | Macroexpand last sexp        |
 | ~SPC m m f~ | Macroexpand surrounding sexp |
-| ~SPC m m r~ | Macroexpand region           |
+| ~SPC m m x~ | Macroexpand region           |
```
Ah yes, strange. I think this was an accident...

dalanicolai replied:
Ah it was also `x` in packages.el... then this must have been a brain shortcut :grin:

dalanicolai commented on layers/+lang/scheme/README.org:
```diff
@@ -183,15 +184,41 @@ When enabled the symbol =🆂= should be displayed in the mode-line.
+| Key binding     | Description           |
+|-----------------+-----------------------|
+| ~S-RET~         | Insert newline indent |
+| ~C-l / M-RET C~ | Clear buffer          |
```
I indeed wanted to remove that (as mentioned in https://github.com/syl20bnr/spacemacs/pull/14758#discussion_r635149752, which I forgot to submit), but it is no typo; `M-RET` in insert mode is the equivalent of `SPC m` in normal mode

dalanicolai commented:
I have updated the PR. As I expect you prefer to remove the `customization shortcut` I have removed it (I have no strong opinion about it). If you like it back then I can of course simply put it back...

lebensterben commented on layers/+lang/scheme/README.org:
```diff
@@ -183,15 +184,40 @@ When enabled the symbol =🆂= should be displayed in the mode-line.
+| ~SPC m k~   | REPL interrupt                                    |
+| ~SPC m q~   | REPL exit                                         |
+
+** Doc-mode
```
'Geiser-doc-mode' is clearer.

lebensterben commented on layers/+lang/scheme/README.org:
```diff
@@ -183,15 +184,40 @@ When enabled the symbol =🆂= should be displayed in the mode-line.
+*Insert state*
+| Key binding | Description           |
+|-------------+-----------------------|
+| ~S-RET~     | Insert newline indent |
```
'Insert newline and indent' or simply 'Insert newline'

lebensterben commented on layers/+lang/scheme/README.org:
```diff
@@ -183,15 +184,40 @@ When enabled the symbol =🆂= should be displayed in the mode-line.
+*Normal state*
+| Key binding | Description                                       |
+|-------------+---------------------------------------------------|
+| ~g j~       | Goto next prompt                                  |
```
you may combine them
|~g j~ / ~]]~ | Go to next prompt     |
|~g k~ / ~[[~ | Go to previous prompt |

lebensterben commented on layers/+lang/scheme/README.org:
```diff
@@ -183,15 +184,40 @@ When enabled the symbol =🆂= should be displayed in the mode-line.
+| ~g k~       | Goto previous prompt                              |
+| ~]]/[[~     | Goto next/previous prompt                         |
+
+| ~SPC m h d~ | Show documentation for symbol at point            |
```
if 'SPC m h h' is not already used, it's better since easier to type

lebensterben commented on layers/+lang/scheme/README.org:
```diff
@@ -183,15 +184,40 @@ When enabled the symbol =🆂= should be displayed in the mode-line.
+|-------------+-----------------------|
+| ~o~         | Follow link           |
+| ~]]/[[~     | Next/previous-section |
+| ~>/<~       | Next/previous page    |
```
```txt
| ~>~ / ~g n~ | Next page     |
| ~<~ / ~g p~ | Previous page |
```

lebensterben commented on layers/+lang/scheme/README.org:
```diff
@@ -183,15 +184,40 @@ When enabled the symbol =🆂= should be displayed in the mode-line.
+| ~g d~       | Goto definition       |
+| ~g z~       | Switch to repl        |
+| ~TAB/S-TAB~ | Next/previous button  |
+| ~C-j/k~     | Next/previous button  |
```
`~C-j/k~` is ambiguous
```txt
| ~TAB~ / ~C-j~   | Next button     |
| ~S-TAB~ / ~C-k~ | Previous button |
```

dalanicolai commented:
Nice feedback! I agree with all of them and I have updated the PR...

lebensterben commented:
merging now. thanks for your contributions.
